### ISSUE TITLE
Add /listitems command

### ISSUE THREAD

**Issue by Wuzzy2** [minetest/minetest#11154]
Basically, I suggest to add a command `/listitems` to list all itemstrings. And maybe also a command `/listentities` to list all entitystrings.

I also suggest to allow filtering by name.

Basically, I suggest to add a simplified version of the `listitems` mod into builtin:
https://forum.minetest.net/viewtopic.php?t=18049

I think the mod in question has a little feature overload, that's why I'm proposing only a simplified version.

## Why into builtin?

Because I think this is a feature you would *always* want, regardless of game. Itemstrings and entitystrings are required arguments for `/give`, `/giveme` and `/spawnentity`. The system of itemstrings and entitystrings is a core feature, not specific to any game of mod. Therefore, it makes sense to add these commands into builtin. There is no good reason IMO *not* to have such a feature available.

## Asking for opinions
What do you think about the idea in general? Good or bad? If bad, it's easy: Just close this issue. But if yes, I might look into how to convert the `listitems` mod to builtin.

**Comment by sfan5** [comment]
I disagree, it should be on games to provide a list of all items to the user _or_ choose not to do so.

`/give`, `/giveme` should be considered debug commands these days. `/spawnentity` is just for debugging.
If you know the technical strings to put into them that's fine, if not then you better use the games' provided way to get items for free.

**Comment by Wuzzy2** [comment]
Well but that's precisely my point. To make it easier to look up the itemstrings for debugging/testing/whatever purposes. Some games can become very complex with thausands of items. It's easy to forget them even if you're the main developer from said game (trust me, I know what I'm talking about!).
Also note that for most games it is rare to provide a truly full list of items. Some items are “internal”. See `/listitems` as a natural extension to `/give`.

If I apply your logic, we should also drop `/give` and `/giveme` because it is “up to the game”.

Oh, I should probably add that these commands would require privileges (probably `give`) to work. No free information for players. ;-)

**Comment by Desour** [comment]
I don't think it's nice to print *all* itemstrings into the chat. The output of `/mods` is already nearly to large.
With filter options however, such command might be useful.

But the real fix to make commands like `/give` easier to use would be autocompletion, imo.

**Comment by Andrey2470T** [comment]
I don't think this feature would be useful for /give or /giveme commands. As well if even /mods _already_ outputs enough often a whole mess of mods names with no possibility to find out something, then what output would that command do? A number of itemstrings will be more _in times_ than a number of mods. I would prefer a name prompt maybe with using 'TAB' key when you are printing a name of itemstring in /give or /giveme.

**Comment by wsor4035** [comment]
I agree, (tab) autocompletion would be ideal

**Comment by voxel-minetest** [comment]
> I don't think it's nice to print _all_ itemstrings into the chat. The output of `/mods` is already nearly to large.
The problem is, that you can not run basic terminal commands like `clear` or maybe just split the actual prompt into two, one is chat, the other is command / debugging stuff only. 
> [...] autocompletion [...] 
That's the other missing basic terminal feature.

I would welcome a list*foobar* command and not just a strip down version of the listitems mods. 
Imo this mod is already a bare necessity.

**Comment by sfan5** [comment]
> Oh, I should probably add that these commands would require privileges (probably `give`) to work. No free information for players.

I assumed that wouldn't be the case, it's not that bad then.

Perhaps it'd be better to add a `-l` / `--list` to `/give` and `/spawnentity` than additional commands.

**Comment by voxel-minetest** [comment]
I found out, that I couldn't scroll back to the start of the output of `/listitems` in vanilla mtg. 
So if you implement such function a higher scrollback value per default would be necessary too.

**Comment by Wuzzy2** [comment]
Adding a flag sounds like a nice idea.

**Comment by AntumDeluge** [comment]
Just a note: I updated the *listitems* mod to no longer support ***/list&lt;type&gt;*** format (e.g. ***/listitems***, ***/listnodes***, etc.). Only ***/list &lt;type&gt;*** is now supported.

The only setting is now ***listitems.bullet_list*** to enable/disable displaying items in a bulleted list.

I agree that built-in functionality would be a good idea.

**Comment by HybridDog** [comment]
Here's another listitems command implementation: https://github.com/red-001/more_chatcommands/blob/4faf1c5e5a710ad907cebc76c7ccd1419f31ae4d/init.lua#L135-L153

**Comment by AntumDeluge** [comment]
Sorry that I keep going back & forth with this, but I think I'm going to change back to the *list&lt;type&gt;* format in my mod. It's more simple. @Wuzzy2 was right about it.

**Comment by voxel-minetest** [comment]
> Sorry that I keep going back & forth with this, but I think I'm going to change back to the _**/list<type>**_ (single word command) format in my mod. It's more simple. @Wuzzy2 was right about it.

Huh? 
Nah, it's way better now!
Only if a `/list` will be in the engine then it would be better to went back, or you would have to do ugly things like `//foobar`.

**Comment by sorcerykid** [comment]
I have mixed opinions about this command. While it's true that recalling itemstrings and entity names can be a pain, I don't think that dumping a gigantic list to chat is very helpful nor user-friendly.

For example, on the Just Test Tribute server, the command `/craft default:*`, shows all craftable items that are registered to the Default mod within an easy to navigate GUI. And, of course, entering `/craft` with no parameters, simply shows all craftable items. At the bottom of the formspec is a field to further refine the search.

Perhaps the `/spawnentity` and `/giveme` commands, could allow the user to select from a scrolling list of matching entities or nodes in a similar way. Hence, Issuing the command `/giveme default:*_diamond` would open a formspec showing all registered diamond tools from the Default mod. There would also be a field to edit the search string. If a quantity is provided (in the case of craftitems or nodes), then that would be preserved once the selection is made. Just a thought.
